core/src/replay_stage.rs

```diff
@@ -673,6 +686,57 @@ impl ReplayStage {
         Ok(())
     }
 
+    fn snapshot_tower(tower_snapshot_path: &PathBuf, tower: &Tower) -> Result<()> {
+        let timer = Instant::now();
+        fs::create_dir_all(tower_snapshot_path)?;
+        let mut snapshot_file = File::create(tower_snapshot_path.join(TOWER_SNAPSHOT_NAME))?;
+        snapshot_file.write_all(&bincode::serialize(&tower).expect("tower serialize failed"))?;
+        snapshot_file.flush()?;
+        let snapshot_time = timer.elapsed().as_millis() as usize;
+        inc_new_counter_info!("replay_stage-tower_snapshot_duration_ms", snapshot_time);
+        Ok(())
+    }
+
+    fn reload_tower(
+        my_pubkey: &Pubkey,
+        tower_snapshot_path: &PathBuf,
+        vote_account: &Pubkey,
+        bank_forks: &Arc<RwLock<BankForks>>,
+    ) -> Tower {
+        info!("Looking for Tower snapshot at {:?}", tower_snapshot_path);
+        // try to reload tower from disk, otherwise reconstruct one
+        let tower =
+            File::open(tower_snapshot_path.join(TOWER_SNAPSHOT_NAME)).and_then(|tower_file| {
+                let mut stream = BufReader::new(tower_file);
+                let tower: bincode::Result<Tower> = bincode::deserialize_from(&mut stream);
+                tower.map_err(|err| io::Error::new(io::ErrorKind::Other, format!("{:?}", err)))
+            });
+        // check that the tower actually belongs to this node
+        match tower {
+            Ok(tower) => {
+                if &tower.node_pubkey != my_pubkey {
+                    error!(
+                        "Wrong tower state found. My pubkey {:?} but found tower for {:?}",
+                        my_pubkey, tower.node_pubkey
+                    );
+                    Tower::new(&my_pubkey, &vote_account, &bank_forks.read().unwrap())
+                } else {
+                    info!("Restoring tower from saved state..");
+                    tower
+                }
+            }
+            Err(e) => {
+                if e.kind() != std::io::ErrorKind::NotFound {
+                    error!(
+                            "Error: {:?} Unable to restore tower, generating a new one from from bank forks",
+                            e
+                        );
+                }
+                Tower::new(&my_pubkey, &vote_account, &bank_forks.read().unwrap())
```

Comment:
Tower::new() will try to initialize the heaviest bank in bank_forks as root. I'm not sure this is the correct thing to do if we can't find tower state/tower state is corrupted. 

Thinking we should panic, ask for manual intervention (probably use private key to transfer funds out into a different account to avoid slashing).

@aeyakovenko, what do you think?

Reply:
Snapshot recovery, basically needs to send a tx, assert we are on the current fork, then expire max censorship threshold

Reply:
What is max censorship threshold?

Reply:
@aeyakovenko can you elaborate on what you mean by "expire max censorship threshold"?

Reply:
@TristanDebrunner wait for 2^threshold slots to start voting.  This is necessary if the VoteState is reused between reboots.  If it’s a new VoteState there is no chance of accidentally voting on a fork.

Reply:
I assume you're referring to the threshold introduced in #6744?

Reply:
It's part of the current implementation already.

https://github.com/solana-labs/solana/blob/master/core/src/consensus.rs#L19

Reply:
How about changing this to return a `Result<Tower>`? Then if it returns `Err`, replay stage can check if the validator's vote account exists in the bank. If it does exist and isn't empty, we panic as @carllin suggested. Otherwise, it's a new validator, so we can use the default that currently gets returned silently.

Then once the relevant checks in #6936 are implemented, replay stage can wait instead of `panic`ing

Reply:
Should we also double check the signature to make sure the state wasn't corrupted?

Reply:
@TristanDebrunner - is @carllin's question resolved?  I see you asked for review again 7 days ago